turbo/rpchelper/subscription.go

```diff
@@ -0,0 +1,132 @@
+package rpchelper
+
+import (
+	"context"
+	"sync"
+)
+
+// a simple interface for subscriptions for rpc helper
+type Sub[T any] interface {
+	Send(T)
+	Close()
+}
+
+type chan_sub[T any] struct {
+	ch chan T
+
+	closed chan struct{}
+
+	ctx context.Context
+	cn  context.CancelFunc
+
+	c sync.Cond
+}
+
+// buffered channel
+func newChanSub[T any](size int) *chan_sub[T] {
+	// set min size to 8.
+	if size < 8 {
+		size = 8
+	}
+	o := &chan_sub[T]{}
+	o.ch = make(chan T, size)
+	o.closed = make(chan struct{})
+	o.ctx, o.cn = context.WithCancel(context.Background())
+	return o
+}
+func (s *chan_sub[T]) Send(x T) {
+	select {
+	// if the output buffer is empty, send
+	case s.ch <- x:
+		// if sub is canceled, dispose message
+	case <-s.ctx.Done():
+		// the sub is overloaded, dispose message
+	default:
+	}
+}
+func (s *chan_sub[T]) Close() {
+	select {
+	case <-s.ctx.Done():
+		return
+	default:
+	}
+	// its possible for multiple goroutines to get to this point, if Close is called twice at the same time
+	// close the context - allows any sends to exit
+	s.cn()
+	select {
+	case s.closed <- struct{}{}:
+		// but it is not possible for multiple goroutines to get to this point
+		// drain the channel
+		for range s.ch {
+		}
+		close(s.ch)
+	default:
+	}
+}
+
+func NewSyncMap[K comparable, T any]() *SyncMap[K, T] {
+	return &SyncMap[K, T]{
+		m: make(map[K]T),
+	}
+}
+
+type SyncMap[K comparable, T any] struct {
+	m map[K]T
+	sync.RWMutex
+}
+
+func (m *SyncMap[K, T]) Get(k K) (res T, ok bool) {
+	m.RLock()
+	defer m.RUnlock()
+	res, ok = m.m[k]
+	return res, ok
+}
+
+func (m *SyncMap[K, T]) Put(k K, v T) (T, bool) {
+	m.Lock()
+	defer m.Unlock()
+	_, ok := m.m[k]
+	m.m[k] = v
+	return m.m[k], ok
```

Comment:
usually more useful to return old value instead of new. For example: then user can implement some metric like `size = size + len(v) - len(oldV)`

Reply:
done